lib/motions/general-motions.coffee

```diff
@@ -104,15 +104,51 @@ class Motion
 class CurrentSelection extends Motion
   constructor: (@editor, @vimState) ->
     super(@editor, @vimState)
-    @selection = @editor.getSelectedBufferRanges()
+    @lastSelection = @editor.getSelectedBufferRange()
+    @wasLinewise = @isLinewise()
 
   execute: (count=1) ->
     _.times(count, -> true)
 
   select: (count=1) ->
-    @editor.setSelectedBufferRanges(@selection)
+    # in visual mode, the current selections are already there
+    # if we're not in visual mode, we are repeating some operation and need to re-do the selections
+    unless @vimState.mode is 'visual'
+      if @wasLinewise
+        @selectLines()
+      else
+        @selectCharacters()
+
     _.times(count, -> true)
 
+  selectLines: ->
+    {start, end} = @lastSelection
+    lineCount = end.row - start.row
+    for selection in @editor.getSelections()
+      cursor = selection.cursor.getBufferPosition()
+      selection.setBufferRange
+        start:
+          row: cursor.row
+          column: 0
+        end:
+          row: cursor.row + end.row - start.row
+          column: 0
+
+  selectCharacters: ->
+    {start, end} = @lastSelection
+    if @lastSelection.isSingleLine()
+      count = end.column - start.column
+      wrap = settings.wrapLeftRightMotion()
+      for selection in @editor.getSelections()
+        _.times count, ->
+          selection.selectRight() if wrap or not selection.cursor.isAtEndOfLine()
+    else
+      for selection in @editor.getSelections()
+        cursor = selection.cursor.getBufferPosition()
+        selection.selectToBufferPosition
+          row: cursor.row + end.row - start.row
+          column: end.column
```

Comment:
You might be able to use some more abstract methods on [`Point`](https://github.com/atom/text-buffer/blob/master/src/point.coffee#L144) and [`Range`](https://github.com/atom/text-buffer/blob/master/src/range.coffee#L300) to do this. Could you try something along these lines?

```coffee
selectCharacters: ->
  lastSelectionExtent = @lastSelection.getExtent()
  for selection in @editor.getSelections()
    {start} = selection.getBufferRange()
    selection.setBufferRange([start, start.traverse(lastSelectionExtent)])
```